packages/lesswrong/lib/instanceSettings.ts

```diff
@@ -0,0 +1,113 @@
+import { initializeSetting } from './publicSettings'
+import { Meteor } from 'meteor/meteor'
+
+const getNestedProperty = function (obj, desc) {
+  var arr = desc.split('.');
+  while(arr.length && (obj = obj[arr.shift()]));
+  return obj;
+};
+
+export const Settings: Record<string,any> = {};
+
+// Keep track of which settings have been used (`getSetting`) without being
+// registered (`registerSetting`). Registering settings is optional, but if a
+// setting is registered with a default value after it has already been used
+// without a default value, that's an error.
+const settingsUsedWithoutRegistration: Record<string,boolean> = {};
+
+const getSetting = <T>(settingName: string, settingDefault?: T): T => {
+
+  let setting;
+
+  // if a default value has been registered using registerSetting, use it
+  if (typeof settingDefault === 'undefined' && Settings[settingName])
+    settingDefault = Settings[settingName].defaultValue;
+
+  // If this setting hasn't been registered, and is used here without a default
+  // value specified, record the fact that it was used without registration.
+  if (!(settingName in Settings) && (typeof settingDefault === 'undefined')) {
+    settingsUsedWithoutRegistration[settingName] = true;
+  }
+
+  if (Meteor.isServer) {
+    // look in public, private, and root
+    const rootSetting = getNestedProperty(Meteor.settings, settingName);
+    const privateSetting = Meteor.settings.private && getNestedProperty(Meteor.settings.private, settingName);
+    const publicSetting = Meteor.settings.public && getNestedProperty(Meteor.settings.public, settingName);
+    
+    // if setting is an object, "collect" properties from all three places
+    if (typeof rootSetting === 'object' || typeof privateSetting === 'object' || typeof publicSetting === 'object') {
+      setting = {
+        ...settingDefault,
+        ...rootSetting,
+        ...privateSetting,
+        ...publicSetting,
+      };
+    } else {
+      if (typeof rootSetting !== 'undefined') {
+        setting = rootSetting;
+      } else if (typeof privateSetting !== 'undefined') {
+        setting = privateSetting;
+      } else if (typeof publicSetting !== 'undefined') {
+        setting = publicSetting;
+      } else {
+        setting = settingDefault;
+      }
+    }
+
+  } else {
+    // look only in public
+    const publicSetting = Meteor.settings.public && getNestedProperty(Meteor.settings.public, settingName);
+    setting = typeof publicSetting !== 'undefined' ? publicSetting : settingDefault;
+  }
+
+  // Settings[settingName] = {...Settings[settingName], settingValue: setting};
+
+  return setting;
+
+};
+
+export class PublicInstanceSetting<SettingValueType> {
```

Comment:
Needs a comment -- the fact that "instance" meant "uses meteor's config system" wasn't obvious.

```suggestion
// A setting which is configured via Meteor's configuration system, ie in `settings.json`, as opposed to a database setting.
export class PublicInstanceSetting<SettingValueType> {
```